src/github.com/getlantern/errors/errors.go

```diff
@@ -0,0 +1,641 @@
+  }
+
+If you want additional logging, call WithLogging(true) during program bootstrapping. If that flag is set, Report() will get a logger using golog.LoggerFor("<package-name-in-which-the-error-is-created">) and call its Error() method.
+
+It's the caller's responsibility to avoid race condition accessing same error instance from multiple goroutines.
+*/
+package errors
+
+import (
+	"bufio"
+	"bytes"
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/hex"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net"
+	"net/http"
+	"net/textproto"
+	"net/url"
+	"os"
+	"os/exec"
+	"reflect"
+	"runtime"
+	"strconv"
+	"syscall"
+	"time"
+
+	"github.com/getlantern/golog"
+	"github.com/getlantern/jibber_jabber"
+	"github.com/getlantern/osversion"
+	"github.com/getlantern/stack"
+)
+
+var (
+	defaultSystemInfo *systemInfo
+	currentReporter   Reporter = &StdReporter{}
+	logging                    = false
+)
+
+func init() {
+	osVersion, _ := osversion.GetHumanReadable()
+	defaultSystemInfo = &systemInfo{
+		OSType:    runtime.GOOS,
+		OSArch:    runtime.GOARCH,
+		OSVersion: osVersion,
+		GoVersion: runtime.Version(),
+		//TODO: app vesion is also required. As a low level library, we need it to be initialized by caller.
+	}
+}
+
+type Reporter interface {
+	Report(*Error)
+}
+
+type StdReporter struct{}
+
+func (l StdReporter) Report(e *Error) {
+	fmt.Printf("%+v", e.Error())
+}
+
+func ReportTo(r Reporter) {
+	currentReporter = r
+}
+
+func WithLogging(b bool) {
+	logging = b
+}
+
+func New(s string) (e *Error) {
+	e = &Error{
+		GoType: "errors.Error",
+		Desc:   s,
+	}
+	e.attachStack(2)
+	return
+}
+
+func Wrap(err error) (e *Error) {
+	return WrapSkipFrame(err, 1)
+}
+
+func WrapSkipFrame(err error, skip int) (e *Error) {
+	if e, ok := err.(*Error); ok {
+		return e
+	}
+	e = &Error{Source: err}
+	// always skip [WrapSkipFrame, attachStack]
+	e.attachStack(skip + 2)
+	e.applyDefaults()
+	return
+}
+
+// Error wraps system and application errors in unified structure
+type Error struct {
+	// Source captures the underlying error that's wrapped by this Error
+	Source  error           `json:"-"`
+	Stack   stack.CallStack `json:"-"`
+	Package string          `json:"package"` // lantern
+	Func    string          `json:"func"`    // foo.Bar
+	// FileLine is the file path relative to GOPATH together with the line when
+	// Error is created.
+	FileLine string `json:"file_line"` // github.com/lantern/foo.go:10
+	// ReportFileLine is the file and line where the error is reported
+	ReportFileLine string `json:"report_file_line"`
+	// Go type name or constant/variable name of the error
+	GoType string `json:"type"`
+	// Error description, by either Go library or application
+	Desc string `json:"desc"`
+	// The operation which triggers the error to happen
+	Op string `json:"operation,omitempty"`
+	// Any extra fields
+	Extra map[string]string `json:"extra,omitempty"`
+
+	*ProxyingInfo
+	*UserLocale
+	*UserAgentInfo
+}
+
+func (e *Error) Report() {
+	caller := stack.Caller(1)
```

Comment:
A function call is inevitable to `applyDefaults` and attach the stack, and should at the very place the Error is created (to get the accurate stack). It seems like we should avoid creating Error manually.